File: php/40-combination-sum2.php
<?php
class Solution {
    private $target;
    private $count;
    private $data = [];
    /**
     * @param Integer[] $candidates
     * @param Integer $target
     * @return Integer[][]
     */
    function combinationSum2($candidates, $target) {
        $this->target = $target;
        $this->count = count($candidates);
        sort($candidates);
        $this->helper($candidates, 0, []);
        return $this->data;
    }
    function helper(&$candidates, $start, $val) {
        $sum = array_sum($val);
        if ($sum > $this->target) {
            return;
        }
        if ($sum == $this->target) {
            $this->data[] = $val;
            return;
        }
        for ($i = $start; $i < $this->count; $i++) {
            if ($i != $start && $candidates[$i] == $candidates[$i-1]) {
                continue;
            }
            if ($sum + $candidates[$i] > $this->target) continue;
            array_push($val, $candidates[$i]);
            $this->helper($candidates, $i+1, $val);
            
            array_pop($val);
        }
    }
}

File: go/40-combination-sum2.go
package main

import "sort"

func combinationSum2(candidates []int, target int) [][]int {
	res := [][]int{}
	if len(candidates) == 0 {
		return res
	}
	sort.Ints(candidates)
	start := 0
	helper(candidates, target, start, []int{}, &res)

	return res
}
func helper(candidates []int, target int, start int, el []int, res *[][]int) {
	sum := getSum(el)
	if sum > target {
		return
	}
	if sum == target {
		copyEl := make([]int, len(el))
		copy(copyEl, el)
		*res = append(*res, el)
		return
	}
	for i := start; i < len(candidates); i++ {
		if i != start && candidates[i] == candidates[i-1] {
			continue
		}
		if sum+candidates[i] > target {
			continue
		}
		helper(candidates, target, i+1, append(el, candidates[i]), res)
	}
}
func getSum(el []int) int {
	sum := 0
	for _, v := range el {
		sum += v
	}
	return sum
}
